// cm-md-editor/src/MdEditor.js
/**
 * Author: Stefan Haack (https://shaack.com)
 * Date: 2023-11-12
 */
import {defaultTools} from "./tools/DefaultTools.js"

export class MdEditor {

    constructor(element, props) {
        this.element = element
        this.props = {
            colorHeading: "100,160,255",
            colorCode: "130,170,200",
            colorComment: "128,128,128",
            colorLink: "100,180,220",
            colorBlockquote: "100,200,150",
            colorList: "100,200,150",
            colorStrikethrough: "255,100,100",
            colorBold: "255,180,80",
            colorItalic: "180,130,255",
            colorHtmlTag: "200,120,120",
            colorHorizontalRule: "128,128,200",
            colorEscape: "128,128,128",
            colorFrontMatter: "128,128,200",
            wordWrap: true,
            tools: defaultTools,
            ...props
        }
        this.tools = this.props.tools.map(entry => {
            if (Array.isArray(entry)) {
                const [Tool, toolProps] = entry
                return new Tool(this, toolProps)
            }
            return new entry(this)
        })
        this.element.addEventListener('keydown', (e) => this.handleKeyDown(e))
        this.createToolbar()
        this.createHighlightBackdrop()
        if (!this.wrapEnabled && this.highlightLayer) {
            this.highlightLayer.style.whiteSpace = 'pre'
            this.highlightLayer.style.overflowWrap = 'normal'
        }
    }

    createToolbar() {
        const wrapper = document.createElement('div')
        this.element.parentNode.insertBefore(wrapper, this.element)
        wrapper.appendChild(this.element)
        const toolbar = document.createElement('div')
        toolbar.style.cssText = 'display:flex;gap:1px;padding:2px;flex-wrap:wrap;background:rgba(128,128,128,0.15);border:1px solid rgba(128,128,128,0.3);border-bottom:none;border-radius:4px 4px 0 0;box-sizing:border-box;width:100%;'
        wrapper.insertBefore(toolbar, this.element)
        this.element.style.borderRadius = '0 0 4px 4px'
        for (const tool of this.tools) {
            if (typeof tool.toolbarButtons === 'function') {
                tool.toolbarButtons().forEach(btn => this.createToolbarButton(toolbar, btn))
            }
        }

        // Spacer to push wrap toggle to the right
        const spacer = document.createElement('div')
        spacer.style.cssText = 'flex:1;'
        toolbar.appendChild(spacer)

        // Wrap toggle button
        this.wrapStorageKey = 'mdEditor_wrap_' + (this.element.id || this.element.name || 'default')
        const savedWrap = localStorage.getItem(this.wrapStorageKey)
        this.wrapEnabled = savedWrap !== null ? savedWrap === 'true' : this.props.wordWrap
        this.wrapButton = document.createElement('button')
        this.wrapButton.type = 'button'
        this.wrapButton.title = 'Toggle word wrap'
        this.wrapButton.style.cssText = 'background:none;border:none;border-radius:3px;cursor:pointer;padding:4px 6px;display:flex;align-items:center;justify-content:center;color:inherit;opacity:0.6;transition:opacity 0.15s,background 0.15s;'
        this.wrapButton.style.width = '32px'
        this.wrapButton.style.height = '28px'
        this.loadIcon('text-wrap.svg').then(svg => { this.wrapButton.innerHTML = svg })
        this.wrapButton.addEventListener('mouseenter', () => { this.wrapButton.style.opacity = '1'; this.wrapButton.style.background = 'rgba(128,128,128,0.2)' })
        this.wrapButton.addEventListener('mouseleave', () => { this.wrapButton.style.opacity = this.wrapEnabled ? '0.9' : '0.4'; this.wrapButton.style.background = 'none' })
        this.wrapButton.addEventListener('mousedown', (e) => e.preventDefault())
        this.wrapButton.addEventListener('click', (e) => {
            e.preventDefault()
            this.toggleWrapMode()
        })
        this.wrapButton.style.opacity = this.wrapEnabled ? '0.9' : '0.4'
        toolbar.appendChild(this.wrapButton)
        // Apply saved wrap state
        if (!this.wrapEnabled) {
            this.element.style.whiteSpace = 'pre'
            this.element.style.overflowX = 'auto'
        }
    }

    createToolbarButton(toolbar, btn) {
        if (btn.separator) {
            const sep = document.createElement('div')
            sep.style.cssText = 'width:1px;align-self:stretch;margin:3px 5px;background:rgba(128,128,128,0.4);'
            toolbar.appendChild(sep)
            return
        }
        if (btn.element) {
            toolbar.appendChild(btn.element)
            return
        }
        const button = document.createElement('button')
        button.type = 'button'
        button.title = btn.title
        button.style.cssText = 'background:none;border:none;border-radius:3px;cursor:pointer;padding:4px 6px;display:flex;align-items:center;justify-content:center;color:inherit;opacity:0.6;transition:opacity 0.15s,background 0.15s;'
        if (btn.icon) {
            button.innerHTML = `<svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 16 16" fill="currentColor">${btn.icon}</svg>`
        } else if (btn.iconUrl) {
            button.style.width = '32px'
            button.style.height = '28px'
            this.loadIconFromUrl(btn.iconUrl).then(svg => { button.innerHTML = svg })
        } else if (btn.iconFile) {
            button.style.width = '32px'
            button.style.height = '28px'
            this.loadIcon(btn.iconFile).then(svg => { button.innerHTML = svg })
        }
        button.addEventListener('mouseenter', () => { button.style.opacity = '1'; button.style.background = 'rgba(128,128,128,0.2)' })
        button.addEventListener('mouseleave', () => { button.style.opacity = '0.6'; button.style.background = 'none' })
        button.addEventListener('mousedown', (e) => e.preventDefault())
        button.addEventListener('click', (e) => {
            e.preventDefault()
            btn.action()
        })
        toolbar.appendChild(button)
    }

    loadIcon(filename) {
        const baseUrl = new URL('./tools/icons/', import.meta.url)
        return this.loadIconFromUrl(new URL(filename, baseUrl).href)
    }

    loadIconFromUrl(url) {
        return fetch(url)
            .then(r => r.text())
            .then(svg => svg.replace(/width="16"/, 'width="20"').replace(/height="16"/, 'height="20"'))
    }

    createHighlightBackdrop() {
        const container = this.element.parentNode
        container.style.position = 'relative'
        this.backdrop = document.createElement('div')
        this.highlightLayer = document.createElement('div')
        this.backdrop.appendChild(this.highlightLayer)
        container.appendChild(this.backdrop)

        // Copy textarea computed styles to backdrop
        const cs = window.getComputedStyle(this.element)
        this.backdrop.style.cssText = `position:absolute;overflow:hidden;pointer-events:none;z-index:1;`
        this.highlightLayer.style.cssText = `white-space:pre-wrap;word-wrap:break-word;overflow-wrap:break-word;`

        const syncStyles = () => {
            const cs = window.getComputedStyle(this.element)
            const props = ['font-family', 'font-size', 'font-weight', 'line-height', 'letter-spacing',
                'tab-size', 'padding-top', 'padding-right', 'padding-bottom', 'padding-left']
            props.forEach(p => this.highlightLayer.style[p.replace(/-([a-z])/g, (_, c) => c.toUpperCase())] = cs.getPropertyValue(p))
            this.highlightLayer.style.boxSizing = 'border-box'
            // Use clientWidth to match the textarea's content area (excludes scrollbar)
            this.highlightLayer.style.width = this.element.clientWidth + 'px'
            const borderTop = parseInt(cs.getPropertyValue('border-top-width')) || 0
            const borderLeft = parseInt(cs.getPropertyValue('border-left-width')) || 0
            this.backdrop.style.top = (this.element.offsetTop + borderTop) + 'px'
            this.backdrop.style.left = (this.element.offsetLeft + borderLeft) + 'px'
            this.backdrop.style.width = this.element.clientWidth + 'px'
            this.backdrop.style.height = this.element.clientHeight + 'px'
        }
        syncStyles()

        // Move textarea bg to backdrop, make textarea transparent so backdrop shows through
        const originalBg = cs.getPropertyValue('background-color')
        this.backdrop.style.background = originalBg
        this.element.style.overscrollBehavior = 'none'
        this.element.style.background = 'transparent'
        this.element.style.position = 'relative'
        this.element.style.zIndex = '2'
        const originalColor = cs.color
        this.element.style.caretColor = originalColor
        this.element.style.color = 'transparent'
        this.highlightLayer.style.color = originalColor

        // Sync scroll via transform — no visible jitter since all text is in the highlight layer
        const syncScroll = () => {
            this.highlightLayer.style.transform = `translate(${-this.element.scrollLeft}px, ${-this.element.scrollTop}px)`
        }
        this.element.addEventListener('scroll', syncScroll)

        // Update on input
        this.element.addEventListener('input', () => this.updateHighlight())
        new ResizeObserver(() => { syncStyles(); this.updateHighlight() }).observe(this.element)
        this.updateHighlight()
    }

    colorSpan(colorProp, content) {
        return '<span style="color:rgba(' + this.props[colorProp] + ',1)">' + content + '</span>'
    }

    updateHighlight() {
        const text = this.element.value
        const lines = text.split('\n')
        let html = ''
        let inCodeBlock = false
        let inHtmlComment = false
        let inFrontMatter = false
        // YAML front matter must start at the very first line
        if (lines.length > 0 && lines[0].trim() === '---') {
            inFrontMatter = true
        }

        for (let i = 0; i < lines.length; i++) {
            if (i > 0) html += '\n'
            const line = lines[i]

            // YAML front matter
            if (inFrontMatter) {
                html += this.colorSpan('colorFrontMatter', this.escapeHtml(line))
                if (i > 0 && line.trim() === '---') {
                    inFrontMatter = false
                }
                continue
            }

            // Fenced code block delimiter
            if (/^`{3,}/.test(line)) {
                inCodeBlock = !inCodeBlock
                html += this.colorSpan('colorCode', this.escapeHtml(line))
                continue
            }
            if (inCodeBlock) {
                html += this.colorSpan('colorCode', this.escapeHtml(line))
                continue
            }

            // HTML comment handling (can span multiple lines)
            if (inHtmlComment) {
                const endIdx = line.indexOf('-->')
                if (endIdx !== -1) {
                    inHtmlComment = false
                    html += this.colorSpan('colorComment', this.escapeHtml(line.substring(0, endIdx + 3)))
                    html += this.highlightInline(line.substring(endIdx + 3))
                } else {
                    html += this.colorSpan('colorComment', this.escapeHtml(line))
                }
                continue
            }
            if (line.trimStart().startsWith('<!--')) {
                const endIdx = line.indexOf('-->', line.indexOf('<!--') + 4)
                if (endIdx !== -1) {
                    // Single-line comment
                    html += this.colorSpan('colorComment', this.escapeHtml(line))
                } else {
                    // Multi-line comment starts
                    inHtmlComment = true
                    html += this.colorSpan('colorComment', this.escapeHtml(line))
                }
                continue
            }

            // Horizontal rule (3+ of same -, *, or _ with optional spaces)
            if (/^\s{0,3}([-*_])\s*(\1\s*){2,}$/.test(line)) {
                html += this.colorSpan('colorHorizontalRule', this.escapeHtml(line))
                continue
            }

            // Headings
            const headingMatch = line.match(/^(#{1,6}) /)
            if (headingMatch) {
                const opacity = Math.max(0.3, 1 - (headingMatch[1].length - 1) * 0.1)
                html += '<span style="color:rgba(' + this.props.colorHeading + ',' + opacity + ')">' + this.escapeHtml(line) + '</span>'
                continue
            }

            // Reference link definition [ref]: url
            const refMatch = line.match(/^(\s{0,3}\[)([^\]]+)(\]:\s+)(.+)$/)
            if (refMatch) {
                html += this.colorSpan('colorLink', this.escapeHtml(refMatch[1]))
                    + this.colorSpan('colorLink', this.escapeHtml(refMatch[2]))
                    + this.colorSpan('colorLink', this.escapeHtml(refMatch[3]))
                    + this.colorSpan('colorLink', this.escapeHtml(refMatch[4]))
                continue
            }

            // Blockquote prefix
            let prefix = ''
            let rest = line
            const bqMatch = line.match(/^(\s*>+\s?)/)
            if (bqMatch) {
                prefix = this.colorSpan('colorBlockquote', this.escapeHtml(bqMatch[0]))
                rest = line.substring(bqMatch[0].length)
            }

            html += prefix + this.highlightInline(rest)
        }
        // Trailing newline so the backdrop height matches the textarea
        this.highlightLayer.innerHTML = html + '\n'
    }

    highlightInline(line) {
        // Split by inline code to protect code content from other highlighting
        const segments = []
        let lastIndex = 0
        const codeRegex = /`([^`]+)`/g
        let match

        while ((match = codeRegex.exec(line)) !== null) {
            if (match.index > lastIndex) {
                segments.push({type: 'text', content: line.substring(lastIndex, match.index)})
            }
            segments.push({type: 'code', content: match[0]})
            lastIndex = codeRegex.lastIndex
        }
        if (lastIndex < line.length) {
            segments.push({type: 'text', content: line.substring(lastIndex)})
        }
        if (segments.length === 0) {
            segments.push({type: 'text', content: ''})
        }

        let result = ''
        for (const seg of segments) {
            if (seg.type === 'code') {
                result += this.colorSpan('colorCode', this.escapeHtml(seg.content))
            } else {
                result += this.highlightTextSegment(this.escapeHtml(seg.content))
            }
        }
        return result
    }

    highlightTextSegment(escaped) {
        let result = escaped
        const c = (prop) => this.props[prop]

        // Escape sequences: dim the backslash before markdown punctuation
        result = result.replace(/\\([\\`*_{}[\]()#+\-.!~|])/g,
            '<span style="color:rgba(' + c('colorEscape') + ',1)">\\</span>$1')

        // Unordered list markers with optional task list checkbox
        result = result.replace(/^((?:\t|  )*)(- )(\[[ xX]\] )?/, (_, tabs, marker, task) => {
            let r = tabs + this.colorSpan('colorList', marker)
            if (task) {
                r += this.colorSpan('colorList', task)
            }
            return r
        })

        // Ordered list markers
        result = result.replace(/^((?:\t|  )*)(\d+\. )/, (_, tabs, marker) =>
            tabs + this.colorSpan('colorList', marker))

        // Images ![alt](url) and Links [text](url)
        result = result.replace(/(!?\[)(.*?)(\]\()(.+?)(\))/g, (_, p1, p2, p3, p4, p5) =>
            this.colorSpan('colorLink', p1) + this.colorSpan('colorLink', p2) + this.colorSpan('colorLink', p3) + this.colorSpan('colorLink', p4) + this.colorSpan('colorLink', p5))

        // Reference links [text][ref]
        result = result.replace(/(\[)(.*?)(\]\[)(.*?)(\])/g, (_, p1, p2, p3, p4, p5) =>
            this.colorSpan('colorLink', p1) + this.colorSpan('colorLink', p2) + this.colorSpan('colorLink', p3) + this.colorSpan('colorLink', p4) + this.colorSpan('colorLink', p5))

        // Strikethrough ~~text~~
        result = result.replace(/(~~)(.*?)(~~)/g, (_, p1, p2, p3) =>
            this.colorSpan('colorStrikethrough', p1) + this.colorSpan('colorStrikethrough', p2) + this.colorSpan('colorStrikethrough', p3))

        // Bold **text**
        result = result.replace(/(\*\*)(.*?)(\*\*)/g, (_, p1, p2, p3) =>
            this.colorSpan('colorBold', p1) + this.colorSpan('colorBold', p2) + this.colorSpan('colorBold', p3))

        // Italic _text_ or *text* (single asterisk, after bold has been handled)
        result = result.replace(/((?:^|[^\\*]))(\_)(.*?[^\\])(\_)/g, (_, pre, p1, p2, p3) =>
            pre + this.colorSpan('colorItalic', p1) + this.colorSpan('colorItalic', p2) + this.colorSpan('colorItalic', p3))
        result = result.replace(/((?:^|[^\\*]))(\*)((?!\*).+?[^\\])(\*)/g, (_, pre, p1, p2, p3) =>
            pre + this.colorSpan('colorItalic', p1) + this.colorSpan('colorItalic', p2) + this.colorSpan('colorItalic', p3))

        // HTML tags
        result = result.replace(/(&lt;)(\/?[a-zA-Z]\w*)(.*?)(&gt;)/g, (_, p1, p2, p3, p4) =>
            this.colorSpan('colorHtmlTag', p1 + p2 + p3 + p4))

        // Tool inline highlighting
        for (const tool of this.tools) {
            if (typeof tool.highlightInline === 'function') {
                result = tool.highlightInline(result)
            }
        }

        return result
    }

    toggleWrapMode() {
        this.wrapEnabled = !this.wrapEnabled
        localStorage.setItem(this.wrapStorageKey, this.wrapEnabled)
        const wrap = this.wrapEnabled
        this.element.style.whiteSpace = wrap ? 'pre-wrap' : 'pre'
        this.element.style.overflowX = wrap ? 'hidden' : 'auto'
        this.highlightLayer.style.whiteSpace = wrap ? 'pre-wrap' : 'pre'
        this.highlightLayer.style.overflowWrap = wrap ? 'break-word' : 'normal'
        this.wrapButton.style.opacity = wrap ? '0.9' : '0.4'
        this.updateHighlight()
    }

    escapeHtml(str) {
        return str.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;')
    }

    getCurrentLineInfo() {
        const start = this.element.selectionStart
        const text = this.element.value
        const lineStart = start === 0 ? 0 : text.lastIndexOf('\n', start - 1) + 1
        let lineEnd = text.indexOf('\n', start)
        if (lineEnd === -1) lineEnd = text.length
        const line = text.substring(lineStart, lineEnd)
        return {lineStart, lineEnd, line}
    }

    selectLineRange(lineStart, lineEnd) {
        this.element.selectionStart = lineStart
        this.element.selectionEnd = lineEnd
    }

    toggleWrap(marker) {
        const start = this.element.selectionStart
        const end = this.element.selectionEnd
        const text = this.element.value
        const len = marker.length
        const before = text.substring(start - len, start)
        const after = text.substring(end, end + len)
        if (before === marker && after === marker) {
            // Remove markers, keep selection on the inner text
            this.element.selectionStart = start - len
            this.element.selectionEnd = end + len
            const selected = text.substring(start, end)
            this.insertTextAtCursor(selected)
            this.element.selectionStart = start - len
            this.element.selectionEnd = end - len
        } else if (start !== end) {
            // Wrap selection, keep selection on the inner text
            this.insertTextAtCursor(marker + text.substring(start, end) + marker)
            this.element.selectionStart = start + len
            this.element.selectionEnd = end + len
        } else {
            this.insertTextAtCursor(marker + marker)
            this.element.selectionStart = this.element.selectionEnd = start + len
        }
    }

    insertTextAtCursor(text) {
        // execCommand is deprecated, but without alternative to insert text and preserve the correct undo/redo stack
        document.execCommand("insertText", false, text)
    }

    handleKeyDown(e) {
        const before = this.element.value.substring(0, this.element.selectionStart)
        const currentLine = before.substring(before.lastIndexOf('\n') + 1)
        const isListMode = currentLine.match(/^\t*- /) || currentLine.match(/^\t*\d+\. /)
        if (e.key === 'Tab') {
            e.preventDefault()
            if (isListMode) {
                if (!e.shiftKey) {
                    this.insertTabAtLineStart()
                } else {
                    this.removeTab()
                }
            } else {
                this.insertTabAtCursorPosition()
            }
        } else if (e.key === 'Enter') {
            this.handleEnterKey(e)
        } else if (e.ctrlKey || e.metaKey) {
            for (const tool of this.tools) {
                if (typeof tool.keyboardShortcuts === 'function') {
                    for (const shortcut of tool.keyboardShortcuts()) {
                        if (shortcut.ctrlOrMeta && shortcut.key === e.key) {
                            e.preventDefault()
                            shortcut.action(e)
                            return
                        }
                    }
                }
            }
        }
    }

    handleEnterKey(e) {
        const start = this.element.selectionStart
        const before = this.element.value.substring(0, start)
        const currentLine = before.substring(before.lastIndexOf('\n') + 1)
        const matchEmptyUl = currentLine.match(/^(\s*- )$/)
        const matchEmptyOl = currentLine.match(/^(\s*)\d+\. $/)
        const matchHyphen = currentLine.match(/^(\s*- )/)
        const matchOl = currentLine.match(/^(\s*)(\d+)\. ./)
        if (matchEmptyUl) {
            this.element.selectionStart = this.element.selectionEnd - matchEmptyUl[1].length - 1
        } else if (matchEmptyOl) {
            this.element.selectionStart = this.element.selectionEnd - matchEmptyOl[0].length - 1
        } else if (matchHyphen) {
            e.preventDefault()
            this.insertTextAtCursor('\n' + matchHyphen[1])
        } else if (matchOl) {
            e.preventDefault()
            const nextNum = parseInt(matchOl[2]) + 1
            this.insertTextAtCursor('\n' + matchOl[1] + nextNum + '. ')
        }
    }

    insertTabAtCursorPosition() {
        this.insertTextAtCursor('\t')
    }

    insertTabAtLineStart() {
        const start = this.element.selectionStart
        const before = this.element.value.substring(0, start)
        const lineStart = before.lastIndexOf('\n') + 1
        this.element.selectionStart = this.element.selectionEnd = lineStart
        this.insertTextAtCursor('\t')
        this.element.selectionStart = this.element.selectionEnd = start + 1
    }

    // test

    removeTab() {
        const start = this.element.selectionStart
        const before = this.element.value.substring(0, start)
        const lineStart = before.lastIndexOf('\n') + 1
        const currentLine = before.substring(lineStart)
        if (currentLine.startsWith('\t')) {
            this.element.selectionStart = lineStart
            this.element.selectionEnd = lineStart + 1
            this.insertTextAtCursor("")
            this.element.selectionStart = this.element.selectionEnd = start - 1
        }
    }
}
